src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_ExpressionTrees.cs

```diff
@@ -291,18 +291,29 @@ public override BoundNode VisitCompoundAssignmentOperator(BoundCompoundAssignmen
 
         public override BoundNode Visit(BoundNode node)
         {
-            if (_inExpressionLambda &&
-                // Ignoring BoundConversion nodes prevents redundant diagnostics
-                !(node is BoundConversion) &&
-                node is BoundExpression expr &&
-                expr.Type is TypeSymbol type &&
-                type.IsRestrictedType())
+            if (_inExpressionLambda && node is BoundExpression expression)
             {
-                Error(ErrorCode.ERR_ExpressionTreeCantContainRefStruct, node, type.Name);
+                ReportSuppressionIfNeeded(expression);
+
+                if (!(node is BoundConversion) && // Ignoring BoundConversion nodes prevents redundant diagnostics
+                    expression.Type is TypeSymbol type &&
+                    type.IsRestrictedType())
+                {
+                    Error(ErrorCode.ERR_ExpressionTreeCantContainRefStruct, node, type.Name);
+                }
             }
+
             return base.Visit(node);
         }
 
+        private void ReportSuppressionIfNeeded(BoundExpression expression)
+        {
+            if (expression.IsSuppressed)
+            {
+                Error(ErrorCode.ERR_ExpressionTreeCantContainSuppressNullableWarning, expression);
```

Comment:
>ERR_ExpressionTreeCantContainSuppressNullableWarning [](start = 32, length = 52)

Why disallow `!` in an expression tree?

Won't we report a nullability warning for `=> x`?
```C#
    static void F(object? x)
    {
        Expression<Func<object>> e = () => x;
    }
```

Reply:
When suppressions were represented as a bound node, I think it made more sense to disallow. Now I could go either way, but I'm still leaning on disallowing since it serves no purpose. That reserves our options in the future.

Why do you think we should allow?

---
In reply to: [251513004](https://github.com/dotnet/roslyn/pull/32850#discussion_r251513004) [](ancestors = 251513004)